src/engine/GroupBy.cpp

```diff
@@ -853,52 +837,63 @@ void GroupBy::extractValues(
 }
 
 // _____________________________________________________________________________
-template <size_t numAggregates>
 sparqlExpression::VectorWithMemoryLimit<ValueId>
 GroupBy::extractValuesDirectlyFromMap(
     size_t beginIndex, size_t endIndex,
-    const ad_utility::HashMapWithMemoryLimit<KeyType, ValueType<numAggregates>>&
-        map,
+    const ad_utility::HashMapWithMemoryLimit<KeyType, ValueType>& map,
+    const std::vector<std::vector<AverageAggregationData>>& aggregationData,
     IdTable* resultTable, size_t hashMapIndex, size_t outCol) {
   // Store in a vector such that we can copy it to
   // `_previousResultsFromSameGroup`.
   sparqlExpression::VectorWithMemoryLimit<ValueId> aggregateResults(
       getExecutionContext()->getAllocator());
 
+  aggregateResults.resize(endIndex - beginIndex);
+
   // Store aggregate results in table and in vector.
   decltype(auto) groupValues = resultTable->getColumn(0);
-  for (size_t j = beginIndex; j < endIndex; j++) {
-    Id val = groupValues[j];
-    auto& aggregateData = map.at(val).at(hashMapIndex);
-    auto aggregateResult = aggregateData.calculateResult();
-    (*resultTable)(j, outCol) = aggregateResult;
-    aggregateResults.push_back(aggregateResult);
-  }
+  decltype(auto) outValues = resultTable->getColumn(outCol);
+
+  auto op = [&map, &hashMapIndex, &aggregationData](Id val) {
+    auto aggregateData = aggregationData.at(hashMapIndex).at(map.at(val));
+    return aggregateData.calculateResult();
+  };
+
+  std::ranges::transform(groupValues.begin() + beginIndex,
+                         groupValues.begin() + endIndex,
```

Comment:
Something else that I find here:
Don't you have exactly the same code in the case where you in addition have to compute an expression on the result?
So the only difference is that here you do an additional copy to the output directly, whereas in the other case you first pass it through the expression module?
I am sure here we can get rid of some redundancy.